crates/prql-ast/src/expr/literal.rs

```diff
@@ -20,3 +20,62 @@ pub struct ValueAndUnit {
     pub n: i64,       // Do any DBs use floats or decimals for this?
     pub unit: String, // Could be an enum IntervalType,
 }
+
+impl std::fmt::Display for Literal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Literal::Null => write!(f, "null")?,
+            Literal::Integer(i) => write!(f, "{i}")?,
+            Literal::Float(i) => write!(f, "{i}")?,
+
+            Literal::String(s) => {
+                quote_string(s, f)?;
+            }
+
+            Literal::Boolean(b) => {
+                f.write_str(if *b { "true" } else { "false" })?;
+            }
+
+            Literal::Date(inner) | Literal::Time(inner) | Literal::Timestamp(inner) => {
+                write!(f, "@{inner}")?;
+            }
+
+            Literal::ValueAndUnit(i) => {
+                write!(f, "{}{}", i.n, i.unit)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn quote_string(s: &str, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    let s = escape_all_except_quotes(s);
+
+    if !s.contains('"') {
+        return write!(f, "\"{s}\"");
+    }
+
+    if !s.contains('\'') {
+        return write!(f, "'{s}'");
+    }
+
+    // when string contains both single and double quotes
+    // find minimum number of double quotes
+    let mut quotes = "\"\"\"".to_string();
+    while s.contains(&quotes) {
+        quotes += "\"";
+    }
+    write!(f, "{quotes}{s}{quotes}")
```

Comment:
Because we're only adding one quote for every three in `s`, does this work correctly?

Reply:
This is basically encasing s into number of quotes that is the largest number of quotes in s + 1.

So if s is `hello"""""""world`, there are 7 quotes, and this would be encased into 8 quotes.

Reply:
Sorry, that's obvious now. I was misreading `s` vs. `quotes`...